### DISCUSSION TITLE
Drives can have the same WWN's

### DISCUSSION THREAD

**Discussion by Panzer1119** [AnalogJ/scrutiny#914] [Issue Triage]
**Describe the bug**
WWNs aren't guaranteed to be unique. If two drives have the same WWN only one of them gets displayed on the dashboard.
It seems to be not that uncommon for drive manufacturers to use the same WWN in a batch.

**Expected behavior**
Identify drives uniquely and handle drives with the same WWN as if the WWN is empty i.e. falling back to the serial number (or maybe generate an identifier consisting of WWN and serial number).

**Screenshots**
Here you can see that two different SSDs have the same WWN.

![image](https://user-images.githubusercontent.com/23016343/215148020-27db29ac-4290-40fe-8e44-5f91b996ec4a.png)

**Log Files**
Look at `/dev/sdg` and `/dev/sdh` when their WWNs get generated.

https://gist.github.com/Panzer1119/f43cd3a828375c70929f382a27bc2a7b

**Comment by AnalogJ** [comment]
🤦 ugh. WWN is supposed to be partially seeded by the serial number, which would still guarantee unique WWN's in this case, however I've also seen issues where only the first X characters of the serial number are used for WWN entropy -- which would be the case here. 


However, on the same system, WWN is supposed to be unique, as thats how the hardware is able to uniquely identify the device. I'm extremely confused, are you running `smartctl --all /dev/sd*` on different hosts?

**Comment by Panzer1119** [comment]
I thought the WWN is stored on the disk, both drives have the same WWN on another host too.

**Comment by AnalogJ** [comment]
This really does break Scrutiny's identification algorithm. I'm kind of at a loss for what to do here.

**Comment by Panzer1119** [comment]
That's understandable, it's really annoying that the vendors don't care more about this.

One possibility would be to add location information (like physical port etc.) to the identifier, but that isn't really good either, because that can change if the drive is moved.

One thing that I think might work is to make some kind of configuration (file) where you manually define drives that have the same identifier and assign them additional info for identification, like a numeric index (e.g. based on the physical port).

Scrutiny could use UUIDs internally to identify drives if it doesn't already, so the config file could be a manual definition of a UUID for a drive on a physical port.

And maybe Scrutiny could detect if different drives on different physical ports(`/dev/sda`, `/dev/sdb` etc.) have the same identifiers and print a warning in the console and a link to how to handle them.

**Comment by flopon** [comment]
Same issue here.
A lot of identical Samsung SSD in the same NAS, and two of them sharing the same WWN.
Anyone having an idea on how to deal with that ?

**Comment by briandotdev** [comment]
Just found my way here with same issue.  Two SSDs with same WWN, only one shows to the UI.  Any success with a work around or updates on a fix?  

If I follow the steps to reference the disk by-id will that fix this?  When I browse `/dev/disk/by-id` i can see that the references for the WWN are definitely mixed up there too but there are also references to the ata id which do appear to be unique and map to the correct device.

**Comment by AnalogJ** [comment]
We could possibly allow the user to override the WWN to the disk-id instead. We just need to figure out if theres a way to actually get the disk-id programmatically -- and make sure it works on windows & linux :(

**Comment by JacksonLegge** [comment]
I've got the same issue here with a [pair of 2TB Patriot P220 SSDs](https://www.amazon.ca/dp/B0DFMS998B) that came in a 2-pack. Has anyone figured out a workaround for this?

**Comment by boeremak** [comment]
I am running into the same issue with 2 Samsung SSD's.

Would using the drive serial number work? Seems unique and works even when you switch them around in different hosts. (I have multiple machines).

**Comment by aprosnyak** [comment]
I’m experiencing the same issue — both Samsung SSDs have unique serial numbers but share the same WWN identifier.

**Comment by flopon** [comment]
Yes, I'm still affected by the issue, please consider it.

**Comment by kaysond** [comment]
I think the solution may be to switch/fallback to serial numbers as those should be unique too. Thoughts?

**Reply by flopon** [reply]
Yes, it can be automatic , or configurable with a parameter, or use a combination of SN and WWN.

@Panzer1119 has well explained the issue, and possible workarounds, but @AnalogJ seemed to struggle to find the proper solution.

**Reply by boeremak** [reply]
What workarounds?

**Comment by kaysond** [comment]
Tracking here: https://github.com/AnalogJ/scrutiny/issues/923

**Comment by kaysond** [comment]
Hi all. I've implemented a fix for this in the branch here: https://github.com/AnalogJ/scrutiny/tree/scrutiny_uuid

It identifies drives using a uuid that is derived from a combination of wwn, serial number, and model. There's a database migration involved, so I'd appreciate it if a few people could test it out before I merge it. (Make sure to backup your sqlite and influxdb data!) The migration can take a couple minutes.

In a nutshell, you'll clone the repo to that branch (`git clone -b scrutiny_uuid https://github.com/AnalogJ/scrutiny.git`) then build the container (`make docker-web`), and point to the new dev image in docker run/compose (`ghcr.io/analogj/scrutiny-dev:web`).

Feel free to reach out if you're having any issues, and please let me know once you've confirmed things are working.

**Comment by kaysond** [comment]
FYI this is merged into the nightly
